File: Maximum value of difference of a pair of elements and their Index - GFG/maximum-value-of-difference-of-a-pair-of-elements-and-their-index.cpp
//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++

class Solution {
  public:
    int maxValue(int arr[], int N) {
        // code here
        int temp1, temp2;
        int max1 = INT_MIN, max2 = INT_MIN;
        int min1 = INT_MAX, min2 = INT_MAX;
        for (int i = 0; i < N; i++) {
            temp1 = arr[i] + i;
            temp2 = arr[i] - i;
            max1 = max(max1, temp1);
            min1 = min(min1, temp1);
            max2 = max(max2, temp2);
            min2 = min(min2, temp2);
        }
        
        return max((max1 - min1), (max2 - min2));
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int N;
        cin>>N;
        
        int arr[N];
        
        for(int i=0; i<N; i++)
            cin>>arr[i];

        Solution ob;
        cout << ob.maxValue(arr,N) << endl;
    }
    return 0;
}
// } Driver Code Ends

File: Maximum value of difference of a pair of elements and their Index - GFG/maximum-value-of-difference-of-a-pair-of-elements-and-their-index.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader read =
            new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while (t-- > 0) {
            int N = Integer.parseInt(read.readLine());
            
            String S1[] = read.readLine().split(" ");
            
            int[] arr = new int[N];
            
            for(int i=0; i<N; i++)
            {
                arr[i] = Integer.parseInt(S1[i]);
            }

            Solution ob = new Solution();
            System.out.println(ob.maxValue(arr,N));
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution {
    static int maxValue(int[] arr, int N) {
        // code here
        
        // arr[i] + i
        int max1 = Integer.MIN_VALUE;
        int min1 = Integer.MAX_VALUE;
        
        for (int i = 0; i < N; i++) {
            int val = i + arr[i];
            
            max1 = Math.max(max1, val);
            min1 = Math.min(min1, val);
        }
        
        int ans1 = max1 - min1;
        
        
        // i - arr[i]
        int max2 = Integer.MIN_VALUE;
        int min2 = Integer.MAX_VALUE;
        
        for (int i = 0; i < N; i++) {
            int val = i - arr[i];
            
            max2 = Math.max(max2, val);
            min2 = Math.min(min2, val);
        }
        
        int ans2 = max2 - min2;
        
        return Math.max(ans1, ans2);
    }
};